src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/PackagesConfigToPackageReferenceMigrator.cs

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.PackageManagement.Telemetry;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+using NuGet.ProjectManagement.Projects;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.PackageManagement.UI
+{
+    internal static class PackagesConfigToPackageReferenceMigrator
+    {
+        internal static async Task<string> DoUpgradeAsync(
+            INuGetUIContext context,
+            INuGetUI uiService,
+            NuGetProject nuGetProject,
+            IEnumerable<NuGetProjectUpgradeDependencyItem> upgradeDependencyItems,
+            IEnumerable<PackageIdentity> notFoundPackages,
+            bool collapseDependencies,
+            IProgress<ProgressDialogData> progress,
+            CancellationToken token)
+        {
+
+            var startTime = DateTimeOffset.Now;
+            var packagesCount = 0;
+            var status = NuGetOperationStatus.Succeeded;
+
+
+            using (var telemetry = new TelemetryActivity(Guid.Empty))
+            {
+                try
+                {
+                    // 1. Backup files (csproj and packages.config) that will change
+                    var solutionManager = context.SolutionManager;
+                    var msBuildNuGetProject = (MSBuildNuGetProject)nuGetProject;
+                    var msBuildNuGetProjectSystem = msBuildNuGetProject.ProjectSystem;
+                    var backupPath = string.Empty;
+                    try
+                    {
+                        backupPath = CreateBackup(msBuildNuGetProject, solutionManager.SolutionDirectory);
+                    }
+                    catch(Exception ex)
+                    {
+                        status = NuGetOperationStatus.Failed;
+                        // log error message
+                        uiService.ShowError(ex);
+                        uiService.ProjectContext.Log(MessageLevel.Info,
+                            string.Format(CultureInfo.CurrentCulture, Resources.Upgrader_BackupFailed));
+
+                        return null;
+                    }
+                    
+
+                    // 2. Uninstall all packages currently in packages.config
+                    var progressData = new ProgressDialogData(Resources.NuGetUpgrade_WaitMessage, Resources.NuGetUpgrade_Progress_Uninstalling);
+                    progress.Report(progressData);
+
+                    // Don't uninstall packages we couldn't find - that will just fail
+                    var actions = upgradeDependencyItems.Select(d => d.Package)
```

Comment:
By not found, is it packages that were not fully restored?

If that's the case, maybe we should do an up to date check before running the upgrader. 

I don't understand why this isn't a fail fast situation.

Reply:
it's because the nupkgs need to exist only for validation purpose  - it has nothing to do with the fact that the restore will fail/succeed.

Reply:
You already run a restore prior to upgrading actually, so how would that be possible?

Wouldn't you end up with dangling entries ni packages.config?